akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFutureFlowSpec.scala

```diff
@@ -0,0 +1,411 @@
+        .empty[Int]
+        .viaMat {
+          Flow.futureFlow(prFlow.future)
+        }(Keep.right)
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      fNotUsed.value should be(empty)
+      fSeq.value should be(empty)
+
+      prFlow.success {
+        Flow[Int].orElse(Source.single(99))
+      }
+
+      fNotUsed.futureValue should be(NotUsed)
+      fSeq.futureValue should equal(99 :: Nil)
+    }
+
+    "fails properly on materialization failure with a completed future" in assertAllStagesStopped {
+      val (fNotUsed, fSeq) = src10()
+        .viaMat {
+          Flow.futureFlow {
+            Future.successful(Flow[Int].mapMaterializedValue[NotUsed](_ => throw TE("BBOM!")))
+          }
+        }(Keep.right)
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      fNotUsed.failed.futureValue should be(a[NeverMaterializedException])
+      fNotUsed.failed.futureValue.getCause should equal(TE("BBOM!"))
+      fSeq.failed.futureValue should equal(TE("BBOM!"))
+    }
+
+    "fails properly on materialization failure with a late future" in assertAllStagesStopped {
+      val prFlow = Promise[Flow[Int, Int, NotUsed]]
+      val (fNotUsed, fSeq) = src10()
+        .viaMat {
+          Flow.futureFlow(prFlow.future)
+        }(Keep.right)
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      fNotUsed.value should be(empty)
+      fSeq.value should be(empty)
+
+      prFlow.success(Flow[Int].mapMaterializedValue[NotUsed](_ => throw TE("BBOM!")))
+
+      fNotUsed.failed.futureValue should be(a[NeverMaterializedException])
+      fNotUsed.failed.futureValue.getCause should equal(TE("BBOM!"))
+      fSeq.failed.futureValue should equal(TE("BBOM!"))
+    }
+
+    "propagate flow failures with a completed future" in assertAllStagesStopped {
+      val (fNotUsed, fSeq) = src10()
+        .viaMat {
+          Flow.futureFlow {
+            Future.successful {
+              Flow[Int].map {
+                case 5 => throw TE("fail on 5")
+                case x => x
+              }
+            }
+          }
+        }(Keep.right)
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      fNotUsed.futureValue should be(NotUsed)
+      fSeq.failed.futureValue should equal(TE("fail on 5"))
+    }
+
+    "propagate flow failures with a late future" in assertAllStagesStopped {
+      val prFlow = Promise[Flow[Int, Int, NotUsed]]
+      val (fNotUsed, fSeq) = src10()
+        .viaMat {
+          Flow.futureFlow(prFlow.future)
+        }(Keep.right)
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      fNotUsed.value should be(empty)
+      fSeq.value should be(empty)
+
+      prFlow.success {
+        Flow[Int].map {
+          case 5 => throw TE("fail on 5")
+          case x => x
+        }
+      }
+
+      fNotUsed.futureValue should be(NotUsed)
+      fSeq.failed.futureValue should equal(TE("fail on 5"))
+    }
+
+    "allow flow to handle downstream completion with a completed future" in assertAllStagesStopped {
```

Comment:
Is there a reason that there's no "late future" variant for this one?

Reply:
right, I'll add it.